From 42fd964e0e6d468b37fe3c785b57b8f589b71af3 Mon Sep 17 00:00:00 2001
From: Serguei Katkov <serguei.i.katkov@intel.com>
Date: Tue, 29 Oct 2013 12:12:47 +0700
Subject: Dalvik: Iget_Iput_Removal should replace VR at the start of BB as well

BZ: 148031

Iget_Iput_Removal optimization replaces all uses of the hoisting MIR.
However the utility function stops on the Phi node. As a result
if hoisting instruction reaches end of the loop the uses in the
beginning of the loop should be replaced as well.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-ME
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: I646c31654dc433c7886418677e323089a1409a6c
Orig-MCG-Change-Id: I5796001c6aebe37af3ab166e6ef5ec3c9cab7e09
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/InvariantRemoval.cpp |   14 ++++++++++++++
 1 files changed, 14 insertions(+), 0 deletions(-)

diff --git a/vm/compiler/InvariantRemoval.cpp b/vm/compiler/InvariantRemoval.cpp
index f667ef3..8851ff9 100644
--- a/vm/compiler/InvariantRemoval.cpp
+++ b/vm/compiler/InvariantRemoval.cpp
@@ -756,6 +756,20 @@ static void sinkIput (CompilationUnit *cUnit, LoopInformation *info, MIR *mir, M
 
         if (needMove == true)
         {
+            //Now if there is a Phi node for our oldReg we should update its uses
+            //with new reg as well
+            MIR *phi = info->getPhiInstruction (cUnit, oldReg);
+
+            if (phi != 0)
+            {
+                rewriteSuccess = dvmCompilerRewriteMirUses (phi, oldReg, scratch);
+            }
+
+            if (rewriteSuccess == false)
+            {
+                return;
+            }
+
             //Since we want to ensure correctness, we make sure that we copy back from rewritten
             //register to the original register at all loop exits. So we can generate moves now.
             bool isWide = (defined->ssaRep->numDefs > 1);
-- 
1.7.4.1

